### DISCUSSION TITLE
Two hop node info

### DISCUSSION THREAD

**Discussion by shrimonmuke0202** [pyg-team/pytorch_geometric#3172] [Q&A]
How to add three-hop away information in a edge_index using the class TwoHop  ?

**Comment by rusty1s** [comment]
You should be able to do this by stacking the `TwoHop` transform twice:
```python
transform = T.Compose([T.TwoHop(), T.TwoHop()])
```

**Reply by wsad1** [reply]
Correct me if I am wrong , but looking at the TwoHop() code it looks like it computes A^2 (square of adjacency matrix). So wouldn't computing two hop twice give return 4 hop neighbors.

In the below example the node `4` is 4-hops hops away from node `0`.
```
data = Data(edge_index = torch.tensor([[0,1,2,3],[1,2,3,4]]))
from torch_geometric.transforms import TwoHop
th = TwoHop()
print(th(th(data)).edge_index) # this returns 4 hop neighbors
```
tensor([[0, 0, 0, 0, 1, 1, 1, 2, 2, 3],
             [1, 2, 3, 4, 2, 3, 4, 3, 4, 4]])

**Reply by rusty1s** [reply]
Yes, you are right. My bad. So it seems that currently one needs to call `spspmm` in [`TwoHop`](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/transforms/two_hop.py#L16) twice to achieve the desired effect. A custom transform should do the job. Alternatively, we can think about a `KHop` transform.

**Comment by shrimonmuke0202** [comment]
When I applied coellased coalesced=False 
AssertionError                            Traceback (most recent call last)
<ipython-input-12-0e1e104695e2> in <module>()
     90         for epoch in range(NUM_EPOCHS):
     91             hidden,cell = model.init_hidden(batch_size=TRAIN_BATCH_SIZE)
---> 92             train(model, device, train_loader, optimizer, epoch+1,hidden,cell)
     93             G,P = predicting(model, device, test_loader,hidden,cell)
     94             ret = [rmse(G,P),mse(G,P),pearson(G,P),spearman(G,P)]

8 frames
/usr/local/lib/python3.7/dist-packages/torch_sparse/storage.py in __init__(self, row, rowptr, col, value, sparse_sizes, rowcount, colptr, colcount, csr2csc, csc2csr, is_sorted)
     68                 assert row.max().item() < sparse_sizes[0]
     69             if col.numel() > 0:
---> 70                 assert col.max().item() < sparse_sizes[1]
     71 
     72         if row is not None:

AssertionError: 

This type of error ocurrus . How can I recover from this error?

**Reply by rusty1s** [reply]
Can you give some more information about when this error happens? Do you have an example to reproduce?

**Reply by shrimonmuke0202** [reply]
AssertionError                            Traceback (most recent call last)
<ipython-input-15-0e1e104695e2> in <module>()
     90         for epoch in range(NUM_EPOCHS):
     91             hidden,cell = model.init_hidden(batch_size=TRAIN_BATCH_SIZE)
---> 92             train(model, device, train_loader, optimizer, epoch+1,hidden,cell)
     93             G,P = predicting(model, device, test_loader,hidden,cell)
     94             ret = [rmse(G,P),mse(G,P),pearson(G,P),spearman(G,P)]

9 frames
<ipython-input-15-0e1e104695e2> in train(model, device, train_loader, optimizer, epoch, hidden, cell)
     18         data = data.to(device)
     19         optimizer.zero_grad()
---> 20         output = model(data,hidden,cell)
     21         loss = loss_fn(output, data.y.view(-1, 1).float().to(device))
     22         loss.backward()

/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py in _call_impl(self, *input, **kwargs)
   1049         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1050                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1051             return forward_call(*input, **kwargs)
   1052         # Do not call functions when jit is used
   1053         full_backward_hooks, non_full_backward_hooks = [], []

<ipython-input-14-ab8c34e0f1f2> in forward(self, data, hidden, cell)
     72       #edge_index = torch.cat([data.edge_index,index],dim=1)
     73       #edge_index_square,_ = coalesce(edge_index, None, data.num_nodes, data.num_nodes)
---> 74       squared_data = self.twohop(data)
     75       h4 = self.Conv1(x,squared_data.edge_index)
     76       h4 = self.relu(h4)

/usr/local/lib/python3.7/dist-packages/torch_geometric/transforms/two_hop.py in __call__(self, data)
     14         value = edge_index.new_ones((edge_index.size(1), ), dtype=torch.float)
     15 
---> 16         index, value = spspmm(edge_index, value, edge_index, value, N, N, N)
     17         value.fill_(0)
     18         index, value = remove_self_loops(index, value)

/usr/local/lib/python3.7/dist-packages/torch_sparse/spspmm.py in spspmm(indexA, valueA, indexB, valueB, m, k, n, coalesced)
     28                      sparse_sizes=(k, n), is_sorted=not coalesced)
     29 
---> 30     C = matmul(A, B)
     31     row, col, value = C.coo()
     32 

/usr/local/lib/python3.7/dist-packages/torch_sparse/matmul.py in matmul(src, other, reduce)
    137         return spmm(src, other, reduce)
    138     elif isinstance(other, SparseTensor):
--> 139         return spspmm(src, other, reduce)
    140     raise ValueError
    141 

/usr/local/lib/python3.7/dist-packages/torch_sparse/matmul.py in spspmm(src, other, reduce)
    114            reduce: str = "sum") -> SparseTensor:
    115     if reduce == 'sum' or reduce == 'add':
--> 116         return spspmm_sum(src, other)
    117     elif reduce == 'mean' or reduce == 'min' or reduce == 'max':
    118         raise NotImplementedError

/usr/local/lib/python3.7/dist-packages/torch_sparse/matmul.py in spspmm_sum(src, other)
    104         valueC = valueC.to(value.dtype)
    105     return SparseTensor(row=None, rowptr=rowptrC, col=colC, value=valueC,
--> 106                         sparse_sizes=(M, K), is_sorted=True)
    107 
    108 

/usr/local/lib/python3.7/dist-packages/torch_sparse/tensor.py in __init__(self, row, rowptr, col, value, sparse_sizes, is_sorted)
     23                                      rowcount=None, colptr=None, colcount=None,
     24                                      csr2csc=None, csc2csr=None,
---> 25                                      is_sorted=is_sorted)
     26 
     27     @classmethod

/usr/local/lib/python3.7/dist-packages/torch_sparse/storage.py in __init__(self, row, rowptr, col, value, sparse_sizes, rowcount, colptr, colcount, csr2csc, csc2csr, is_sorted)
     68                 assert row.max().item() < sparse_sizes[0]
     69             if col.numel() > 0:
---> 70                 assert col.max().item() < sparse_sizes[1]
     71 
     72         if row is not None:

AssertionError: 

when Train my model in a batch this type of error occurs but for batch=1 this type of error does not show to me.

**Reply by rusty1s** [reply]
This seems to be a problem with `spspmm` on CUDA. Can you give me some information about your CUDA version?

**Reply by shrimonmuke0202** [reply]
cuda version 11.2 nvidia tesla v100 gpu

**Comment by shrimonmuke0202** [comment]
I WANT three hop away information from a particular node.

**Reply by rusty1s** [reply]
See my answer above.

**Comment by shrimonmuke0202** [comment]
How to multiply a sparse adjacency matrix with a dense adjacency matrix.

**Reply by rusty1s** [reply]
Either:
```python
from torch_scatter import scatter

row, col = edge_index
out = scatter(x[col], row, dim=0, reduce='add')
```
or
```
from torch_sparse import SparseTensor

adj = SparseTensor(row=edge_index[0], col=edge_index[1])
out = adj @ x
```
